rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+
+```bash
+vast exec 'from vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The beauty is that we can almost write the same without requiring a VAST node
+and execute *in situ*:
+
+```bash
+vast exec 'from - |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The only difference being `-` here.
+
+The VAST node basically becomes a **pipeline manager** with persistent state. We
+can keep everything as is, e.g., start a node:
+
+```bash
+vast start -e 1.2.3.4
+```
+
+There are of course a bunch of convenience things we can do, like assuming `from
+-` and `to -` being the default. This would make the above look like this:
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
+
+Side effect: we would get our envisioned convert operator almost for free:
+
+```bash
+vast exec 'parse zeek | print csv'
+# Obviates the need for:
+vast exec 'convert zeek csv'
+```
+
+Another UX improvement would be assuming that we interact with a VAST node. Here
+we could have dedicate commands the prepend or append a pipeline operator:
+
+- `vast pull` → `from vast://...`
+- `vast push` → `to vast://...`
+
+An `export` would then become:
+
+```bash
+vast pull 'where EXPRESSION | print json'
+# Same with exec:
+vast exec 'from vast://.... | where EXPRESSION | print json'
+# Same with export:
+vast export json EXPRESSION
+```
+
+An `import` degenerates to
+
+```bash
+vast push 'parse zeek | where EXPRESSION'
+# Same with exec
+vast exec 'parse zeek | where EXPRESSION | to vast://...'
+# Same with import
+vast import zeek EXPRESSION
+```
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the middle.
+It's clean and simple.
+
+But this may be too restrictive. Let's assume for a moment that input and output
+are dynamically typed. Then an operator must define its input and output types,
+e.g., `from<In, Out>`. This can come in handy for working with matchers:
+
+Here, we build a matcher, with read having the type `read<Bytes, Arrow>`:
+
+```bash
+vast exec 'read csv | matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+The type of `matcher build` would be `<Arrow, Flatbuffer>`. Also assume `matcher
+load` yields `<Flatbuffer, Void>` where `Void` is "no output, just a side
+effect".
+
+How do we pre-load a matcher?
+
+```bash
+vast exec 'from file:///ips.flatbuffer |
+           matcher load --name=ips vast:///1.2.3.4'
+```
+
+Or more UNIX'ish, assuming an implicit `from -`:
+
+```bash
+vast exec 'matcher load --name=ips' < ips.flatbuffer
+```
+
+The `matcher` sub-command `match` does the actual matching:
+
+```bash
+vast pull 'where #type == "suricata.flow" |
+           matcher match --on=:addr ips'
+```
+
+The above is a locally running pipeline, but most likely I want this to run at a
+node. This begs the challenge: how do we manage all these pipelines? With a
+client command (and ideally also a REST API):
```

Comment:
I've thought previously about adding a generic `vast remote [args...]` command that is equivalent to running `vast [args...]` in a remote VAST process.

In general, I think we should not strive to build further one-offs for this. `vast spawn source` and `vast spawn sink` as remote alternatives to `vast import` and `vast export` respectively are already annoying enough to maintain as-is, and we should learn from that experience.

Reply:
I'm not sure if I see the relationship to the commented part in the proposal. Are you saying that the `pipeline` command should not exist, but rather a generic `remote` command? It almost sounds like an `api` command that maps to a REST interface.

Reply:
I don't understand how the quoted section (even after you rewrote it) relates to pipeline management.

To me, pipeline management is about being able to name pipelines, similar to what we already have in the vast.yaml today, such that I can just run `vast exec my-pipeline` to pull it from the configuration.

In which VAST process I want to execute my pipeline should be independent of the CLI of pipelines, e.g., `vast remote exec my-pipeline`. This is both easier to maintain and easier to teach to users.

Reply:
Ah, now I understand what you are saying. You want to inject a `remote` command that takes the subsequent args to a remote VAST node.

Yes, I agree that `spawn source/sink` are one-offs. There should be a generic command to provide a config that does the needed changes to the state, and the dual would be getting a command to get the state.

Regarding `remote exec`, I think this brings us back to the original problem. When you `vast remote exec` something, it's the equivalent of ephemeral matchers, which we want to avoid. I think it's clearer when we do explicit operations for state management, e.g., `vast remote pipeline create foo PIPELINE`, then `vast remote pipeline run foo`.

Reply:
To me, that's a matter of state and not configuration. You could just say `vast exec --persistent 'foo | bar'` and `vast remote exec --persistent 'foo | bar'`.

What should the semantic of a persisted pipeline be? That to me is unclear yet from the doc.

Reply:
I propose moving all the remote execution and state management into a follow-up proposal and to have this one focus on the architecture of `vast exec` itself and that alone. I think that feature alone has so much use in simplifying VAST and making it easier to use that we should have it.

On the other hand, the notion of "every building block is a pipeline oeprator" should probably be included in this proposal, including how we can implement this with backwards compatibility by having `vast {import,export,...}` map onto `vast exec something`.

I would much appreciate a table with a list of all existing commands, and what their `vast exec` equivalent would be, with the explicit notion that the idea is not to replace the commands, but rather to replace their underlying implementation piece by piece with pipeline operators.

Reply:
I added such a table. This was a really helpful exercise because now I'm seeing what's actually going on: we need a remote equivalent of `exec`, otherwise we can't express pipelines remotely! I called this `spawn` for now.

I don't think we can put remote pipeline execution into a different proposal though. This needs to be sorted out here. Perhaps we could agree that there will be a `vast pipeline` or `vast manage pipeline` command that helps with state management, and that we flesh this thing out separate, but remote vs. local execution must be treated here.